File: typescript/17_threads_and_processes.ts
// ============================================================
// Topic:  Multithreading & Multiprocessing  (FOCUS: worker_threads, child_process)
// Summary: Concurrent execution — shared-memory threads vs separate processes.
// Run:     npx ts-node typescript/17_threads_and_processes.ts
//
// KEY DIFFERENCES:
//   Python:  threading + multiprocessing (stdlib)
//   TS/JS:   worker_threads + child_process (Node.js stdlib)
//   Rust:    std::thread + std::sync::Mutex | std::process::Command
//   Java:    Thread / ExecutorService | ProcessBuilder
//   C++:     std::thread + std::mutex | fork()/exec()
// ============================================================

import { createRequire } from "module";
import * as path from "path";
import { spawnSync } from "child_process";
import {
  Worker,
  isMainThread,
  parentPort,
  workerData,
} from "worker_threads";

const nodeRequire = createRequire(__filename);
const tsconfigPath = path.join(__dirname, "tsconfig.json");
const workerOpts = {
  execArgv: ["-r", nodeRequire.resolve("ts-node/register")],
  env: { ...process.env, TS_NODE_PROJECT: tsconfigPath },
};

// ----- Child process entry (separate OS process) -----
if (process.argv.includes("--child")) {
  console.log("child"); // Output: child
  process.exit(0);
}

// ----- Worker thread entry (same process, separate thread) -----
if (!isMainThread) {
  const job = workerData.job as string;

  // ----- 1. Spawn thread — worker prints -----
  if (job === "spawn") {
    console.log("worker"); // Output: worker
  }
  // ----- 2. Join — empty worker (main prints after first join) -----
  else if (job === "join") {
    /* no-op */
  }
  // ----- 3. Lock — shared counter with mutex -----
  else if (job === "increment") {
    const shared = workerData.shared as SharedArrayBuffer;
    const lock = new Int32Array(shared, 0, 1);
    const counter = new Int32Array(shared, 4, 1);
    for (let i = 0; i < 1000; i++) {
      while (Atomics.compareExchange(lock, 0, 0, 1) !== 0) {
        /* spin */
      }
      Atomics.add(counter, 0, 1);
      Atomics.store(lock, 0, 0);
    }
  }
  // ----- 4. Multiple threads — collect results -----
  else if (job === "workerA") {
    parentPort?.postMessage({ key: "a", value: "A" });
  } else if (job === "workerB") {
    parentPort?.postMessage({ key: "b", value: "B" });
  }
} else {
  function runWorker(data: Record<string, unknown>): Promise<void> {
    return new Promise((resolve, reject) => {
      const w = new Worker(__filename, { workerData: data, ...workerOpts });
      w.on("error", reject);
      w.on("exit", (code) => {
        if (code === 0) resolve();
        else reject(new Error(`worker exited ${code}`));
      });
    });
  }

  function runWorkerMessage<T>(data: Record<string, unknown>): Promise<T> {
    return new Promise((resolve, reject) => {
      const w = new Worker(__filename, { workerData: data, ...workerOpts });
      w.on("message", (msg) => resolve(msg as T));
      w.on("error", reject);
      w.on("exit", (code) => {
        if (code !== 0) reject(new Error(`worker exited ${code}`));
      });
    });
  }

  async function main(): Promise<void> {
    // ----- 1. Spawn thread -----
    await runWorker({ job: "spawn" });

    // ----- 2. Join -----
    console.log("main"); // Output: main
    await runWorker({ job: "join" });
    console.log("joined"); // Output: joined

    // ----- 3. Lock / mutex -----
    const shared = new SharedArrayBuffer(8);
    await Promise.all([
      runWorker({ job: "increment", shared }),
      runWorker({ job: "increment", shared }),
    ]);
    const counter = new Int32Array(shared, 4, 1);
    console.log(`counter: ${counter[0]}`); // Output: counter: 2000

    // ----- 4. Multiple threads -----
    const threadResults: Record<string, string> = {};
    const [msgA, msgB] = await Promise.all([
      runWorkerMessage<{ key: string; value: string }>({ job: "workerA" }),
      runWorkerMessage<{ key: string; value: string }>({ job: "workerB" }),
    ]);
    threadResults[msgA.key] = msgA.value;
    threadResults[msgB.key] = msgB.value;
    console.log(threadResults["a"]); // Output: A
    console.log(threadResults["b"]); // Output: B

    // ----- 5. Separate process -----
    const root = path.resolve(__dirname, "..");
    const result = spawnSync(
      "npx",
      ["ts-node", "typescript/17_threads_and_processes.ts", "--child"],
      { stdio: "inherit", shell: true, cwd: root }
    );

    // ----- 6. Process join -----
    console.log(`exit: ${result.status}`); // Output: exit: 0
  }

  main();
}

// Cross-language note:
//   Python threads share memory (GIL limits CPU parallelism); processes are isolated.
//   worker_threads + SharedArrayBuffer + Atomics map to threading.Lock + shared globals.
//   child_process.spawn is like Python multiprocessing.Process or Java ProcessBuilder.
//   On Windows, guard fork/spawn entry points so child processes do not re-run main logic.

File: python/17_threads_and_processes.py
# ============================================================
# Topic:  Multithreading & Multiprocessing  (FOCUS: threads, locks, processes)
# Summary: Concurrent execution — shared-memory threads vs separate processes.
# Run:     python python/17_threads_and_processes.py
#
# KEY DIFFERENCES:
#   Python:  threading + multiprocessing (stdlib)
#   TS/JS:   Web Workers (no shared memory by default)
#   Rust:    std::thread + std::sync::Mutex | rayon
#   Java:    Thread / ExecutorService | ProcessBuilder
#   C++:     std::thread + std::mutex | fork()/exec()
# ============================================================

import threading
import multiprocessing


# ----- 1. Spawn thread — worker prints -----
def worker_spawn():
    print("worker", flush=True)  # Output: worker


# ----- 2. Join — main waits for thread -----
def worker_join():
    pass


# ----- 3. Lock — shared counter with mutex -----
counter = 0
lock = threading.Lock()


def increment_locked():
    global counter
    for _ in range(1000):
        with lock:
            counter += 1


# ----- 4. Multiple threads — collect results -----
thread_results = {}


def worker_a():
    thread_results["a"] = "A"


def worker_b():
    thread_results["b"] = "B"


# ----- 5 & 6. Separate process + join -----
def child_process():
    print("child", flush=True)  # Output: child


def main():
    # ----- 1. Spawn thread -----
    t = threading.Thread(target=worker_spawn)
    t.start()
    t.join()

    # ----- 2. Join -----
    print("main", flush=True)  # Output: main
    t = threading.Thread(target=worker_join)
    t.start()
    t.join()
    print("joined", flush=True)  # Output: joined

    # ----- 3. Lock / mutex -----
    global counter
    counter = 0
    t1 = threading.Thread(target=increment_locked)
    t2 = threading.Thread(target=increment_locked)
    t1.start()
    t2.start()
    t1.join()
    t2.join()
    print(f"counter: {counter}", flush=True)  # Output: counter: 2000

    # ----- 4. Multiple threads -----
    thread_results.clear()
    t_a = threading.Thread(target=worker_a)
    t_b = threading.Thread(target=worker_b)
    t_a.start()
    t_b.start()
    t_a.join()
    t_b.join()
    print(thread_results["a"], flush=True)  # Output: A
    print(thread_results["b"], flush=True)  # Output: B

    # ----- 5. Separate process -----
    p = multiprocessing.Process(target=child_process)
    p.start()

    # ----- 6. Process join -----
    p.join()
    print(f"exit: {p.exitcode}", flush=True)  # Output: exit: 0


if __name__ == "__main__":
    main()

# Cross-language note:
#   Python threads share memory (GIL limits CPU parallelism); processes are isolated.
#   threading.Lock maps to mutex / synchronized blocks in Java, C++, and Rust.
#   multiprocessing.Process is like Java ProcessBuilder or C fork()+exec().
#   On Windows, guard multiprocessing entry points with if __name__ == "__main__".